File: algorithms/mergesort/merge1.erl
#!/usr/bin/env escript
%% -*- erlang -*-
%%! -smp enable -sname p9
% vim:syn=erlang

-mode(compile).

main(_) ->
    {A,B,C} = erlang:timestamp(),
    random:seed(A,B,C),
    L = [random:uniform(10000)||_<-lists:seq(1,100000)],
    %io:format("Merge sort list: ~p ~n", [L]),
    %io:format("Merge sort list: ~p ~n", [m(L)]).
    m(L),
    io:format("Done").

m([L]) -> [L]; 
m(L)   ->
    {L1,L2} = lists:split(length(L) div 2, L),
    merge(m(L1), m(L2)).

merge(L1, L2)    -> merge(L1, L2, []).
merge([], L2, A) -> A++L2;
merge(L1, [], A) -> A++L1;
merge([H1|T1], [H2|T2], A) when H2>=H1 -> merge(T1, [H2|T2], A++[H1]);
merge([H1|T1], [H2|T2], A) when H1>H2  -> merge([H1|T1], T2, A++[H2]).

File: algorithms/mergesort/merge1.py
#!/usr/bin/python3

import random
from heapq import merge

def m(l):
    if len(l) <= 1: return l
    middle = len(l) // 2
    left, right = l[:middle], l[middle:]
    left  = m(left)
    right = m(right) 

    return list(merge(left, right))

l = [random.randint(1,10000) for _ in range(1,5000000)]
#print(l)
#print(m(l))

m(l)
print("Done")
